src/main/scala/mesosphere/util/state/zk/ZKStore.scala

```diff
@@ -0,0 +1,149 @@
+    */
+  override def load(key: ID): Future[Option[ZKEntity]] = {
+    val node = root(key)
+    require(node.parent == root, "Nested paths are not supported!")
+    node.getData().asScala
+      .map{ data => Some(ZKEntity(node, ZKData(data.bytes), Some(data.stat.getVersion))) }
+      .recover { case ex: NoNodeException => None }
+      .recover(exceptionTransform(s"Could not load key $key"))
+  }
+
+  override def create(key: ID, content: Array[Byte]): Future[ZKEntity] = {
+    val node = root(key)
+    require(node.parent == root, "Nested paths are not supported!")
+    val data = ZKData(key, UUID.randomUUID(), content)
+    node.create(data.toProto.toByteArray).asScala
+      .map(n => ZKEntity(n, data, Some(0))) //first version after create is 0
+      .recover(exceptionTransform(s"Can not create entity $key"))
+  }
+
+  /**
+    * This will store a previously fetched entity.
+    * The entity will be either created or updated, depending on the read state.
+    * @return Some value, if the store operation is successful otherwise None
+    */
+  override def save(entity: PersistentEntity): Future[ZKEntity] = {
+    val zk = zkEntity(entity)
+    def update(version: Int): Future[ZKEntity] = {
+      zk.node.setData(zk.data.toProto.toByteArray, version).asScala
+        .map { data => zk.copy(version = Some(data.stat.getVersion)) }
+        .recover(exceptionTransform(s"Can not update entity $entity"))
+    }
+    def create(): Future[ZKEntity] = {
+      zk.node.create(zk.data.toProto.toByteArray).asScala
+        .flatMap { _.getData().map{ data => zk.copy(version = Some(data.stat.getVersion)) }.asScala }
+        .recover(exceptionTransform(s"Can not update entity $entity"))
+    }
+    zk.version.map(update).getOrElse(create())
+  }
+
+  /**
+    * Delete an entry with given identifier.
+    */
+  override def delete(key: ID): Future[ZKEntity] = {
+    val node = root(key)
+    require(node.parent == root, "Nested paths are not supported!")
+    node.exists().asScala
+      .flatMap { d => node.delete(d.stat.getVersion).asScala.map(n => ZKEntity(n, ZKData(key, UUID.randomUUID()))) }
+      .recover(exceptionTransform(s"Can not delete entity $key"))
+  }
+
+  override def allIds(): Future[Seq[ID]] = {
+    root.getChildren().asScala
+      .map(_.children.map(_.name))
+      .recover(exceptionTransform("Can not list all identifiers"))
+  }
+
+  private def exceptionTransform[T](errorMessage: String): PartialFunction[Throwable, T] = {
+    case ex: KeeperException => throw new StoreCommandFailedException(errorMessage, ex)
+    case ex: Throwable       => throw ex
+  }
+
+  private def zkEntity(entity: PersistentEntity): ZKEntity = {
+    entity match {
+      case zk: ZKEntity => zk
+      case _            => throw new IllegalArgumentException("Can not handle this kind of entity")
+    }
+  }
+
+  private def createPath(path: ZNode): ZNode = {
```

Comment:
This uses blocking which is fine (in my opinion) in initialization but shouldn't be used elsewhere.

Thus I would name this method `createPathBlocking` or something like this.